### PR TITLE
[ir] [bug] Fix demote_operations lacking type_check causing float16

### LINKED ISSUE #1824: [Bug] A simple `i == j` comparision in ndrange-for cause `RuntimeError: float16 not supported`
**Describe the bug**
A simple `i == j` in ndrange-for cause `RuntimeError: float16 not supported`.

**To Reproduce**
```py
import taichi as ti

n = 128

A = ti.field(ti.f32, (n, n))

@ti.kernel
def init():
    for i, j in ti.ndrange(n, n):  # ERROR
    #for i, j in A:  # OK
        if i == j:
            A[i, j] = 1

init()
```

**Log/Screenshots**
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-7i9a9afk
[Taichi] <dev mode>, llvm 10.0.0, commit 7ee19751, linux, python 3.8.3
[Taichi] materializing...
[I 09/03/20 12:30:58.462] [llvm_context.cpp:get_data_type@111] float16
[E 09/03/20 12:30:58.462] [llvm_context.cpp:get_data_type@112] Not supported.


***********************************
* Taichi Compiler Stack Traceback *                                                          
***********************************                                                          
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)                                  
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::TaichiLLVMContext::get_data_type(taichi::lang::DataType)                                                                               
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::visit(taichi::lang::UnaryOpStmt*)                                                                                         
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::visit(taichi::lang::Block*)  
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::visit(taichi::lang::IfStmt*) 
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::visit(taichi::lang::Block*)  
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVMCPU::create_offload_range_for(taichi::lang::OffloadedStmt*)                                                                 
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVMCPU::visit(taichi::lang::OffloadedStmt*)                                                                                    
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::visit(taichi::lang::Block*)  
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::emit_to_module()             
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenLLVM::gen()                        
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::CodeGenCPU::codegen()                     
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::KernelCodeGen::compile()                  
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::Program::compile(taichi::lang::Kernel&)   
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::Kernel::compile()                         
/tmp/taichi-7i9a9afk/taichi_core.so: taichi::lang::Kernel::operator()(taichi::lang::Kernel::LaunchContextBuilder&)                                                                        
/tmp/taichi-7i9a9afk/taichi_core.so(+0x458e04) [0x7fd0e427be04]
/tmp/taichi-7i9a9afk/taichi_core.so(+0x395ff2) [0x7fd0e41b8ff2]
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall
/usr/lib/libpython3.8.so.1.0(+0x13e7e6) [0x7fd0e83937e6]
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/usr/lib/libpython3.8.so.1.0(+0x98f6d) [0x7fd0e82edf6d]
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCallDict
/usr/lib/libpython3.8.so.1.0: _PyObject_Call_Prepend
/usr/lib/libpython3.8.so.1.0(+0x1f5e09) [0x7fd0e844ae09]
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: PyEval_EvalCode
/usr/lib/libpython3.8.so.1.0(+0x1d8248) [0x7fd0e842d248]
/usr/lib/libpython3.8.so.1.0(+0x1d2483) [0x7fd0e8427483]
/usr/lib/libpython3.8.so.1.0: PyRun_FileExFlags
/usr/lib/libpython3.8.so.1.0: PyRun_SimpleFileExFlags
/usr/lib/libpython3.8.so.1.0: Py_RunMain
/usr/lib/libpython3.8.so.1.0: Py_BytesMain
/usr/lib/libc.so.6: __libc_start_main
python(_start+0x2e) [0x55b1ae0e604e]

Internal Error occurred, check this page for possible solutions:
https://taichi.readthedocs.io/en/stable/install.html#troubleshooting
Traceback (most recent call last):
  File "a.py", line 13, in <module>
    init()
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 574, in wrapped
    return primal(*args, **kwargs)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 502, in __call__
    return self.compiled_functions[key](*args)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 460, in func__
    t_kernel(launch_ctx)
RuntimeError: [llvm_context.cpp:get_data_type@112] Not supported.
```

**If you have local commits (e.g. compile fixes before you reproduce the bug), please make sure you first make a PR to fix the build errors and then report the bug.**
This bug seems does not exist in 0.6.28:
```
[bate@archit ~]$ p Develop/taichi/a.py
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-vbiyf7lm
[Taichi] version 0.6.28, llvm 10.0.0, commit a58a9af7, linux, python 3.8.3
[Taichi] materializing...
[bate@archit ~]$ pip install taichi==0.6.30
    ...
    Uninstalling taichi-0.6.28:
      Successfully uninstalled taichi-0.6.28
Successfully installed taichi-0.6.30
[bate@archit ~]$ p Develop/taichi/a.py     
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-pvvsqjb9
[Taichi] version 0.6.30, llvm 10.0.0, commit 6052a541, linux, python 3.8.3
[Taichi] materializing...
[I 09/03/20 12:34:01.428] [llvm_context.cpp:get_data_type@111] float16
[E 09/03/20 12:34:01.428] [llvm_context.cpp:get_data_type@112] Not supported.
```

### LINKED ISSUE #1829: [Bug] Integer division with ndrange-related loop indices no longer works
**To Reproduce**
```python
import taichi as ti

ti.init()

@ti.kernel
def test():
    for i, j in ti.ndrange(10, 10):
        if i // 2 == 0:
            print(i)

test()
```

This gives you "floor/ceil takes real inputs only"

**Log/Screenshots**
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-prqalxg3
[Taichi] <dev mode>, llvm 10.0.0, commit bd9b2bbc, linux, python 3.6.9
[Taichi] Starting on arch=x64
[Taichi] materializing...
[E 09/03/20 10:52:16.994] [type_check.cpp:visit@190] [$81] floor/ceil takes real inputs only. At

***********************************
* Taichi Compiler Stack Traceback *
***********************************
/tmp/taichi-prqalxg3/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::TypeCheck::visit(taichi::lang::UnaryOpStmt*)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::TypeCheck::visit(taichi::lang::Block*)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::OffloadedStmt::all_blocks_accept(taichi::lang::IRVisitor*)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::TypeCheck::visit(taichi::lang::Block*)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::irpass::type_check(taichi::lang::IRNode*)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::irpass::offload_to_executable(taichi::lang::IRNode*, taichi::lang::CompileConfig const&, bool, bool, bool, bool)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::irpass::compile_to_executable(taichi::lang::IRNode*, taichi::lang::CompileConfig const&, bool, bool, bool, bool, bool, bool, bool)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::Kernel::lower(bool)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::Program::compile(taichi::lang::Kernel&)
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::Kernel::compile()
/tmp/taichi-prqalxg3/taichi_core.so: taichi::lang::Kernel::operator()(taichi::lang::Kernel::LaunchContextBuilder&)
/tmp/taichi-prqalxg3/taichi_core.so(+0x4644f4) [0x7fe5248064f4]
/tmp/taichi-prqalxg3/taichi_core.so(+0x3aae8d) [0x7fe52474ce8d]
python3.6(_PyCFunction_FastCallDict+0x154) [0x55efede5bc54]
python3.6(_PyObject_FastCallDict+0x2bf) [0x55efede5c06f]
python3.6(_PyObject_Call_Prepend+0x63) [0x55efede60aa3]
python3.6(PyObject_Call+0x3e) [0x55efede5ba5e]
python3.6(+0x16b371) [0x55efedeb5371]
python3.6(_PyObject_FastCallDict+0x8b) [0x55efede5be3b]
python3.6(+0x199c0e) [0x55efedee3c0e]
python3.6(_PyEval_EvalFrameDefault+0x30a) [0x55efedf0675a]
python3.6(PyEval_EvalCodeEx+0x966) [0x55efededeff6]
python3.6(+0x1957d4) [0x55efededf7d4]
python3.6(PyObject_Call+0x3e) [0x55efede5ba5e]
python3.6(_PyEval_EvalFrameDefault+0x19e7) [0x55efedf07e37]
python3.6(+0x192e66) [0x55efededce66]
python3.6(_PyFunction_FastCallDict+0x3d8) [0x55efedede598]
python3.6(_PyObject_FastCallDict+0x26f) [0x55efede5c01f]
python3.6(_PyObject_Call_Prepend+0x63) [0x55efede60aa3]
python3.6(PyObject_Call+0x3e) [0x55efede5ba5e]
python3.6(+0x16b371) [0x55efedeb5371]
python3.6(PyObject_Call+0x3e) [0x55efede5ba5e]
python3.6(_PyEval_EvalFrameDefault+0x19e7) [0x55efedf07e37]
python3.6(+0x193136) [0x55efededd136]
python3.6(+0x193ed6) [0x55efededded6]
python3.6(+0x199b95) [0x55efedee3b95]
python3.6(_PyEval_EvalFrameDefault+0x30a) [0x55efedf0675a]
python3.6(PyEval_EvalCodeEx+0x329) [0x55efedede9b9]
python3.6(PyEval_EvalCode+0x1c) [0x55efededf75c]
python3.6(+0x215744) [0x55efedf5f744]
python3.6(PyRun_FileExFlags+0xa1) [0x55efedf5fb41]
python3.6(PyRun_SimpleFileExFlags+0x1c3) [0x55efedf5fd43]
python3.6(Py_Main+0x613) [0x55efedf63833]
python3.6(main+0xee) [0x55efede2d88e]
/lib/x86_64-linux-gnu/libc.so.6: __libc_start_main
python3.6(+0x1c3160) [0x55efedf0d160]

Internal Error occurred, check this page for possible solutions:
https://taichi.readthedocs.io/en/stable/install.html#troubleshooting
Traceback (most recent call last):
  File "tmp.py", line 11, in <module>
    test()
  File "/home/yuanming/repos/taichi/python/taichi/lang/kernel.py", line 574, in wrapped
    return primal(*args, **kwargs)
  File "/home/yuanming/repos/taichi/python/taichi/lang/kernel.py", line 502, in __call__
    return self.compiled_functions[key](*args)
  File "/home/yuanming/repos/taichi/python/taichi/lang/kernel.py", line 460, in func__
    t_kernel(launch_ctx)
RuntimeError: [type_check.cpp:visit@190] [$81] floor/ceil takes real inputs only. At 

```

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1771 fix #1824 fix #1829

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
So we actually set `DataType::float16` as enum 0 purposely to reveal possible errors? I like this design!

### CODE REVIEW

**Review by xumingkuan — APPROVED** [review]
LGTM!

**Comment by xumingkuan** [comment]
I think it's better to set `unknown` as enum 0 -- for now `f16` happens to be the first in `data_type.inc.h`, so it's enum 0.

**Comment by archibate** [comment]
> I think it's better to set unknown as enum 0 -- for now f16 happens to be the first in data_type.inc.h, so it's enum 0.

The same to `binary_op.inc.h`, whose enum 0 is `mul`... Let's discuss and fix these issues in a separate PR.

File: python/taichi/misc/image.py [hunk 1]
```diff
@@ -57,6 +57,8 @@ def imresize(img, w, h=None):
         img = img.to_numpy()
     if h is None:
         h = w
+    if (w, h) == img.shape:
```
Comment (yuanming-hu) [inline python/taichi/misc/image.py:60]: Could `img.shape` be something like `(512, 512, 3)`?

File: taichi/transforms/compile_to_offloads.cpp [hunk 2]
```diff
@@ -171,7 +171,7 @@ void offload_to_executable(IRNode *ir,
   irpass::analysis::verify(ir);
 
   irpass::demote_operations(ir);
-  print("Operations demoted");
+  print("Operations Demoted");
```
Comment (yuanming-hu) [inline taichi/transforms/compile_to_offloads.cpp:174]: please revert this change.

**Review by yuanming-hu — COMMENTED** [review]
Thanks. LGTM in general.

**Review by yuanming-hu — APPROVED** [review]
